src/AutoMapper/ExpressionExtensions.cs

```diff
@@ -11,6 +12,38 @@
 
     internal static class ExpressionExtensions
     {
+        public static Expression ForEach(Expression collection, ParameterExpression loopVar, Expression loopContent)
+        {
+            var elementType = loopVar.Type;
+            var enumerableType = typeof(IEnumerable<>).MakeGenericType(elementType);
+            var enumeratorType = typeof(IEnumerator<>).MakeGenericType(elementType);
+
+            var enumeratorVar = Variable(enumeratorType, "enumerator");
+            var getEnumeratorCall = Call(collection, enumerableType.GetMethod("GetEnumerator"));
```

Comment:
Here it would be interesting to not go through IEnumerable (use the pattern), in order to avoid allocating the Enumerator, just use the struct, for lists for example. I'm just not sure it would work without special casing collection types.